cpp/src/io/parquet/predicate_pushdown.cpp

```diff
@@ -0,0 +1,521 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "reader_impl_helpers.hpp"
+
+#include <cudf/ast/detail/expression_transformer.hpp>
+#include <cudf/ast/detail/operators.hpp>
+#include <cudf/ast/expressions.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/utilities/integer_utils.hpp>
+#include <cudf/detail/utilities/vector_factories.hpp>
+#include <cudf/transform.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/mr/device/per_device_resource.hpp>
+
+#include <algorithm>
+#include <list>
+#include <numeric>
+#include <optional>
+
+namespace cudf::io::detail::parquet {
+
+/**
+ * @brief Converts statistics in column chunks to 2 device columns - min, max values.
+ *
+ */
+struct stats_caster {
+  size_type total_row_groups;
+  std::vector<metadata> const& per_file_metadata;
+  host_span<std::vector<size_type> const> row_group_indices;
+  rmm::cuda_stream_view stream;
+
+  template <typename ToType, typename FromType>
+  static ToType targetType(FromType const value)
+  {
+    if constexpr (cudf::is_timestamp<ToType>()) {
+      return static_cast<ToType>(
+        typename ToType::duration{static_cast<typename ToType::rep>(value)});
+    } else if constexpr (std::is_same_v<ToType, string_view>) {
+      return ToType{nullptr, 0};
+    } else {
+      return static_cast<ToType>(value);
+    }
+  }
+  // uses storage type as T
```

Comment:
I'm surprised clang-format doesn't require blank lines between these definitions...
```suggestion

  // uses storage type as T
```